server/src/main/java/io/druid/server/coordinator/CostBalancerStrategy.java

```diff
@@ -19,44 +19,134 @@
+   * in the same queries
+   *
+   * @param segmentA The first DataSegment.
+   * @param segmentB The second DataSegment.
+   *
+   * @return the joint cost of placing the two DataSegments together on one node.
+   */
+  public static double computeJointSegmentsCost(final DataSegment segmentA, final DataSegment segmentB)
   {
-    if (interval1.getStartMillis() > interval2.getEndMillis()) {
-      return interval1.getStartMillis() - interval2.getEndMillis();
-    } else if (interval2.getStartMillis() > interval1.getEndMillis()) {
-      return interval2.getStartMillis() - interval1.getEndMillis();
-    } else {
+    final Interval intervalA = segmentA.getInterval();
+    final Interval intervalB = segmentB.getInterval();
+
+    final double t0 = intervalA.getStartMillis();
+    final double t1 = (intervalA.getEndMillis() - t0) / MILLIS_FACTOR;
+    final double start = (intervalB.getStartMillis() - t0) / MILLIS_FACTOR;
+    final double end = (intervalB.getEndMillis() - t0) / MILLIS_FACTOR;
+
+    // constant cost-multiplier for segments of the same datsource
+    final double multiplier = segmentA.getDataSource().equals(segmentB.getDataSource()) ? 2.0 : 1.0;
+
+    return INV_LAMBDA_SQUARE * intervalCost(t1, start, end) * multiplier;
+  }
+
+  /**
+   * Computes the joint cost of two intervals X = [x_0 = 0, x_1) and Y = [y_0, y_1)
+   *
+   * cost(X, Y) = \int_{x_0}^{x_1} \int_{y_0}^{y_1} e^{-\lambda |x-y|}dxdy $$
+   *
+   * lambda = 1 in this particular implementation
+   *
+   * Other values of lambda can be calculated by multiplying inputs by lambda
+   * and multiplying the result by 1 / lambda ^ 2
+   *
+   * Interval start and end are all relative to x_0.
+   * Therefore this function assumes x_0 = 0, x1 >= 0, and y1 > y0
+   *
+   * @param x1 end of interval X
+   * @param y0 start of interval Y
+   * @param y1 end o interval Y
+   *
+   * @return joint cost of X and Y
+   */
+  public static double intervalCost(double x1, double y0, double y1)
+  {
+    if (x1 == 0 || y1 == y0) {
       return 0;
     }
+
+    if (y0 < 0) {
+      // swap X and Y
+      double tmp = x1;
+      x1 = y1 - y0;
+      y1 = tmp - y0;
+      y0 = -y0;
+    }
+
+    // Y overlaps X
+    if (y0 < x1) {
+      /**
+       * X   [ A )[ B )[ C )   or  [ A )[ B )
+       * Y        [   )                 [   )[ C )
+       *
+       * A could be empty if y0 == 0
+       * C could be empty if y1 == x1
+       *
+       * cost(X, Y) = cost(A, Y) + cost(B, C) + cost(B, B)
+       */
+      final double beta;  // b1 - y0
+      final double gamma; // c1 - y0
+      if (y1 <= x1) {
+        beta = y1 - y0;
+        gamma = x1 - y0;
+      } else {
+        beta = x1 - y0;
+        gamma = y1 - y0;
+      }
+      return intervalCost(y0, y0, y1) +
+             intervalCost(beta, beta, gamma) +
+             // cost of exactly overlapping intervals of size beta
+             2 * (beta + FastMath.exp(-beta) - 1);
```

Comment:
@drcrallen this can't be a constant, since beta depends on interval start / end. And yes, this is the solution to
`\int_0^{\beta} \int_{0}^{\beta} e^{|x-y|}dxdy = 2 \cdot (\beta + e^{-\beta} - 1)`

<img width="256" alt="screen shot 2016-05-16 at 9 00 54 pm" src="https://cloud.githubusercontent.com/assets/815147/15310875/50cf7f7c-1ba9-11e6-9187-44e586e9af21.png">